Setup build

barne856 opened the pull request:
1. Added pyproject.toml (modern standard as opposed to setup.py)
2. Added LICENSE file for MIT license (same as Java CC SDK)
3. Updated readme with build and install instructions
4. Added new github workflow file to build, test, and publish binary distributions (.whl) to GitHub Releases on this repo. This will run each time a push to main occurs (with branch protection on this implies all tests from the "lint and test" workflow have passed during the PR review). We can make this manual if desired.
Note that the release is set to fail if the version number in the .toml file is not incremented (i.e. the release already exists).


@thwllms, Will you be implementing the PyPI deployment? Otherwise I can work on that if you provide the credentials. I'm thinking we can create a new workflow for the main branch with the `workflow_run` condition making the GitHub release a dependency of the PyPI release?

HenryGeorgist commented:
> Q: For the PyPI distribution name, should we use `cc-python-sdk`, `cc_sdk`, or something else? I feel like Python should be implied if the package is in PyPI.

cc-python-sdk

thwllms commented:
> Q: For the PyPI distribution name, should we use `cc-python-sdk`, `cc_sdk`, or something else? I feel like Python should be implied if the package is in PyPI.

Both, IMO.

`cc-python-sdk` makes sense as a PyPI package name (i.e. `pip install cc-python-sdk`). Plenty of packages have "python" and dashes in the name.

But when you actually use the library in code, `cc_sdk` makes sense (i.e. `from cc_sdk import DataThing`)

PEP-8 has this to say about package and module names:
> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

https://peps.python.org/pep-0008/#package-and-module-names

thwllms commented:
> 1. Added pyproject.toml (modern standard as opposed to setup.py)
> 
>     2. Added LICENSE file for MIT license (same as Java CC SDK)
> 
>     3. Updated readme with build and install instructions
> 
>     4. Added new github workflow file to build, test, and publish binary distributions (.whl) to GitHub Releases on this repo. This will run each time a push to main occurs (with branch protection on this implies all tests from the "lint and test" workflow have passed during the PR review). We can make this manual if desired.
>        Note that the release is set to fail if the version number in the .toml file is not incremented (i.e. the release already exists).
> 
> 
> @thwllms, Will you be implementing the PyPI deployment? Otherwise I can work on that if you provide the credentials. I'm thinking we can create a new workflow for the main branch with the `workflow_run` condition making the GitHub release a dependency of the PyPI release?



> 1. Added pyproject.toml (modern standard as opposed to setup.py)
> 
>     2. Added LICENSE file for MIT license (same as Java CC SDK)
> 
>     3. Updated readme with build and install instructions
> 
>     4. Added new github workflow file to build, test, and publish binary distributions (.whl) to GitHub Releases on this repo. This will run each time a push to main occurs (with branch protection on this implies all tests from the "lint and test" workflow have passed during the PR review). We can make this manual if desired.
>        Note that the release is set to fail if the version number in the .toml file is not incremented (i.e. the release already exists).
> 
> 
> @thwllms, Will you be implementing the PyPI deployment? Otherwise I can work on that if you provide the credentials. I'm thinking we can create a new workflow for the main branch with the `workflow_run` condition making the GitHub release a dependency of the PyPI release?

Sounds good. If you've got time, go ahead on the PyPI deployment. I'll send you the credentials.

thwllms commented on .devcontainer/devcontainer.json:
Admittedly I don't totally grok devcontainers yet. Any resources you recommend, @barne856?

barne856 replied:
I'm not an expert myself, but I do have some notes I've collected over time.

[devcontainers.md](https://github.com/USACE/cc-python-sdk/files/11972260/devcontainers.md)

barne856 replied:
You can also specify a custom Dockerfile rather than an image https://code.visualstudio.com/docs/devcontainers/create-dev-container#_dockerfile

thwllms commented on .github/workflows/create_github_release.yml:
```diff
@@ -0,0 +1,69 @@
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.11
```
If we're targeting 3.10 as the current default Anaconda version, should we build with that instead? Guessing it doesn't actually matter,

barne856 replied:
I don't think the version of python used to build matters. The version of the build package may. However, we should probably use the latest version of the package since it should have the latest bugfixes / features.

thwllms commented on .github/workflows/create_github_release.yml:
```diff
@@ -0,0 +1,69 @@
+name: Create Release
```
Has `create_github_release.yml` been tested?

barne856 replied:
No, but I've used a similar one on another repo. I'm not sure how to actually test it without merging...

barne856 replied:
I have tested the build locally and that works

thwllms replied:
> No, but I've used a similar one on another repo. I'm not sure how to actually test it without merging...

I'm never quite sure how to deal with this, either, apart from changing the target branch -- was just curious.

thwllms commented on README.md:
```diff
@@ -1,2 +1,97 @@
-The Python SDK for developing plugins for Cloud Compute
\ No newline at end of file
+
+[![PyPI version](https://badge.fury.io/py/cc_sdk.svg)](https://badge.fury.io/py/cc_sdk)
```
`cc-python-sdk`

thwllms commented on pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+[project]
```
Does this need a `dependencies` section? Or will a `pyproject.toml` build automatically include dependencies from `requirements.txt`?

thwllms replied:
Doesn't seem like it based on my quick little test -- I built the wheel for `cc-python-sdk` and installed it to a fresh environment. Dependencies were not installed.
![image](https://github.com/USACE/cc-python-sdk/assets/5504267/11be4935-c5c2-4a86-8205-c96566f2a395)

thwllms commented on .github/workflows/create_github_release.yml:
```diff
@@ -0,0 +1,69 @@
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+          pip install toml build
```
I was able to build locally by just installing `requirements.txt` and `build` (`toml` not needed?)

barne856 replied:
toml is only need in the CI/CD process to check that the version was incremented

thwllms reviewed (changes requested):
LGTM overall. Only thing I'd really change here is making sure that `pyproject.toml` can reflect the dependencies that the library actually needs, so the end-user of `cc-python-sdk` doesn't need to hunt them down and install manually (`attrs`, `boto3`, others?)

barne856 commented:
> LGTM overall. Only thing I'd really change here is making sure that `pyproject.toml` can reflect the dependencies that the library actually needs, so the end-user of `cc-python-sdk` doesn't need to hunt them down and install manually (`attrs`, `boto3`, others?)

Good catch, I just updated the toml file. I added >= the versions we are testing on. I'm not sure how far back we could go with the dependency versions, but I think both of these libraries have fairly good backwards compatibility so we should be good for later versions at least
